### PR TITLE
 Fix #182: Convert error codes before passing them to QUIC 

### LINKED ISSUE #182: Wrong model for stream error codes
The stream error codes are modelled incorrectly.

Currently, `wtransport` just passes the error code numeric value (in fact, the same `VarInt`) through to the QUIC / HTTP3 layer.

However, this is not correct. WebTransport is supposed to apply the following conversion to the error codes:

```rust
fn webtransport_code_to_http_code(n: u8) -> wtransport::VarInt {
    let first: u64 = 0x52e4a40fa8db;
    let n: u64 = n.into();
    let val: u64 = first + n + (n / 0x1e);
    val.try_into().unwrap()
}
```

This is based on https://github.com/web-platform-tests/wpt/blob/master/webtransport/resources/webtransport-test-helpers.sub.js#L18 - which is based on https://ietf-wg-webtrans.github.io/draft-ietf-webtrans-http3/draft-ietf-webtrans-http3.html#section-4.3.

### PR DESCRIPTION
**Pull request by barafael**

### CODE REVIEW

**Comment by BiagioFesta** [comment]
Thank you for taking care of this!


Is the reverse map necessary when receiving the code *from* stream's peer?

**Comment by barafael** [comment]
I have no idea! Sounds like yes it is?

File: wtransport-proto/src/error.rs [hunk 1]
```diff
@@ -53,6 +53,16 @@ pub enum ErrorCode {
 }
 
 impl ErrorCode {
+    /// Converts this stream [`ErrorCode`] to HTTP/3 error code.
+    /// <https://ietf-wg-webtrans.github.io/draft-ietf-webtrans-http3/draft-ietf-webtrans-http3.html#section-4.3)>
+    pub fn to_http3(self) -> VarInt {
+        let code = self.to_code();
```
Comment (finnbear) [inline wtransport-proto/src/error.rs:59]: Upon reading all relevant, it appears that the conversion is only for WT **application** error codes and not any of the existing `ErrorCode` variants, all of which are in the space of valid HTTP/3 error codes.

This can be seen as follows:
- Most of the variants are in the `h3_error_codes` module and numerically match HTTP/3 error code entries in https://datatracker.ietf.org/doc/html/draft-ietf-quic-http#iana-error-table
- For the variant in the `qpack_error_codes` namespace, it numerically matches an HTTP/3 error code entry in https://www.rfc-editor.org/rfc/rfc9204.html
- For the variants in the `wt_error_codes` module, they numerically match HTTP/3 error code entries in https://ietf-wg-webtrans.github.io/draft-ietf-webtrans-http3/draft-ietf-webtrans-http3.html#name-http-3-error-code-registrat

In other words, if we wanted application error codes, they would work like this:
```rust
enum ErrorCode{
    Datagram,
    NoError,
    ...
    Application(u8),
}

impl ErrorCode {
    pub fn to_code(self) -> VarInt {
        match self {
            ErrorCode::Datagram => h3_error_codes::H3_DATAGRAM_ERROR,
            ErrorCode::NoError => h3_error_codes::H3_NO_ERROR,
            ...
            ErrorCode::Application(n) => to_http3(n),
        }
    }
}
```

But I'm not sure `Applicatinon` would be a useful variant. Instead, I suggest looking into whether `reset(error_code)` should be a `u8` and go through the `to_http3` conversion:
https://github.com/BiagioFesta/wtransport/blob/4fcb949c2c97d9b13b784a1761001fe2b2213e7a/wtransport/src/stream.rs#L83-L93

Comment (finnbear) [reply wtransport-proto/src/error.rs:59]: There is one thing I'm unsure about:
- https://ietf-wg-webtrans.github.io/draft-ietf-webtrans-http3/draft-ietf-webtrans-http3.html#name-resetting-data-streams implies there are 2^32 web transport application errors
- https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-02.html#name-http-3-error-code-registrat implies there are only 2^8 web transport application errors

I assume this is because these are two different versions of the spec.

**Comment by MOZGIII** [comment]
> I have no idea! Sounds like yes it is?

It definitely is

**Comment by MOZGIII** [comment]
Check my implementation at https://github.com/MOZGIII/xwt/blob/ffc8b2cc135acaf7f5229a01775e3da92a22684e/crates/xwt-wtransport/src/lib.rs#L284-L323

Might be relevant for https://github.com/BiagioFesta/wtransport/pull/211#discussion_r1742406771

**Comment by finnbear** [comment]
> Might be relevant for [#211 (comment)](https://github.com/BiagioFesta/wtransport/pull/211#discussion_r1742406771)

Can confirm; you [use the conversion for application error/close codes](https://github.com/MOZGIII/xwt/blob/ffc8b2cc135acaf7f5229a01775e3da92a22684e/crates/xwt-wtransport/src/lib.rs#L166-L167), which is correct. This conversion could be moved into `wtransport`.

**Comment by MOZGIII** [comment]
Let's do it! I'm busy with building the reset support Web part of the xwt now, but feel free to port my code from that PR
